src/qibolab/instruments/unrolling.py

```diff
@@ -39,3 +63,64 @@ def batch_max_readout(sequences, max_measurements):
             batch.append(sequence)
             batch_measurements += nmeasurements
     yield batch
+
+
+def _waveform(sequence: PulseSequence):
+    # TODO: deduplicate pulses
+    # TODO: count Rectangular and delays separately
+    # TODO: check if readout duration is faithful for the readout pulse
+    return sequence.duration
+
+
+def _readout(sequence: PulseSequence):
+    return len(sequence.ro_pulses)
+
+
+def _instructions(sequence: PulseSequence):
+    return len(sequence)
+
+
+@dataclass(frozen=True, order=True)
+class Bounds:
+    """Instument memory limitations proxies."""
+
+    waveforms: int = field(metadata={"count": _waveform})
+    """Waveforms estimated size."""
+    readout: int = field(metadata={"count": _readout})
+    """Number of readouts."""
+    instructions: int = field(metadata={"count": _instructions})
+    """Number of readouts."""
+
+    @classmethod
+    def update(cls, sequence: PulseSequence):
+        up = {}
+        for f in fields(cls):
+            up[f.name] = f.metadata["count"](sequence)
+
+        return cls(**up)
+
+    def __add__(self, other: "Bounds") -> "Bounds":
+        new = {}
+        for (k, x), (_, y) in zip(asdict(self).items(), asdict(other).items()):
+            new[k] = x + y
+
+        return type(self)(**new)
+
+
+def batch(sequences: list[PulseSequence], bounds: Bounds):
+    """Split a list of sequences to batches.
+
+    Takes into account the various limitations throught the mechanics defined in
+    :cls:`Bounds`, and the numerical limitations specified by the `bounds` argument.
+    """
+    counters = Bounds(0, 0, 0)
+    batch = []
+    for sequence in sequences:
+        update_ = Bounds.update(sequence)
+        if counters + update_ > bounds:
+            yield batch
+            counters, batch = update_, [sequence]
+        else:
+            batch.append(sequence)
+            counters += update_
+    yield batch
```

Comment:
@Jacfomg @stavros11: the loop is the usual one, but the various conditions are in AND, such that you can specify all the bounds simultaneously. As soon as one exceeds, the batch is yielded, and the conditions are all reset.

This is just a proposal, in case we agree, we should at least write some tests.

At the moment, only one condition is taken into account. To obtain the former behavior, just set a stupidly high bound (we could even convert to float, since they are just proxies, and use `np.inf` - but setting something like `int(1e9)` should always be sufficient, and we could even name that constant in this module).

Reply:
I was playing with this and `if counters + update_ > bounds:` will only check the 1st value of `bounds` so I tried to define  `__lt__` and  `__ge__` for `Bounds` but it complains. I think I can't do that on a `dataclass` I will keep looking to see if I find a solution.

Reply:
I didn't test, I first wanted to have your opinion about the idea. Let me check!

Reply:
This should work

```python
@dataclass(frozen=True, order=False)
class Bounds:
    """Instument memory limitations proxies."""

    waveforms: int = field(metadata={"count": _waveform})
    """Waveforms estimated size."""
    readout: int = field(metadata={"count": _readout})
    """Number of readouts."""
    instructions: int = field(metadata={"count": _instructions})
    """Instructions estimated size."""

    @classmethod
    def update(cls, sequence: PulseSequence):
        up = {}
        for f in fields(cls):
            up[f.name] = f.metadata["count"](sequence)

        return cls(**up)

    def __add__(self, other: "Bounds") -> "Bounds":
        new = {}
        for (k, x), (_, y) in zip(asdict(self).items(), asdict(other).items()):
            new[k] = x + y

        return type(self)(**new)

    def __lt__(self, other: "Bounds") -> bool:
        return any(
            getattr(self, f.name) < getattr(other, f.name) for f in fields(self)
        )
    
    def __ge__(self, other: "Bounds") -> bool:
        return all(
            getattr(self, f.name) >= getattr(other, f.name) for f in fields(self)
        )
```

Reply:
You're right, the last two should be `False` instead:
```py
[ins] In [2]: from qibolab.unrolling import Bounds
                                                                                                        
[ins] In [3]: Bounds(1, 2, 3) + Bounds(1, 1, 1)
Out[3]: Bounds(waveforms=2, readout=3, instructions=4)
                                                                                                        
[ins] In [4]: Bounds(1, 2, 3) + Bounds(1, 1, 1) > Bounds(0, 0, 0)
Out[4]: True
                                                                                                        
[ins] In [5]: Bounds(1, 2, 3) + Bounds(1, 1, 1) > Bounds(5, 0, 0)
Out[5]: False
                                                                                                        
[ins] In [6]: Bounds(1, 2, 3) + Bounds(1, 1, 1) > Bounds(0, 5, 0)
Out[6]: True
                                                                                                        
[ins] In [7]: Bounds(1, 2, 3) + Bounds(1, 1, 1) > Bounds(0, 0, 5)
Out[7]: True
```

Reply:
Yeah, I liked it. It should work for more limitation we could find in the future as well and as a small thing for now as well.

Reply:
I made it more minimal, since the `__ge__` should be the negation of `__lt__` (that's why all of them can be generated from `__lt__` and `__eq__`, that's what `@total_ordering` is doing).

Moreover, the `order=True` was generating lexicographical operations, but I overlooked their utility since my line of thought was something like "is this greater? no, I will look at the next one; is this greater?", while the correct one is "are they equal? no, is it greater or smaller?" and that's it.
But for `__eq__` the one by one comparison is fully alright, so I kept that one.